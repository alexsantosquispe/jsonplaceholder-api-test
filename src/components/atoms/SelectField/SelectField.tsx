import cn from 'clsx';
import { Controller } from 'react-hook-form';
import { twMerge } from 'tailwind-merge';

import {
  INPUT_FIELD_ERROR,
  INPUT_FIELD_STYLES
} from '../../../constants/styles';
import { ChevronUpDownIcon } from '../../../icons';
import { Option } from '../../../services/api.types';
import { InputProps } from '../../../types';

type SelectFieldProps = Omit<InputProps, 'type'> & {
  options: Option[];
};

export const SelectField = ({
  label,
  name,
  options,
  control,
  placeholder = 'Select an option',
  errorMessage,
  isDisabled,
  isRequired
}: SelectFieldProps) => {
  const defaultValueOption = {
    value: '',
    label: placeholder
  } satisfies Option;
  const optionsWithPlaceholder = [defaultValueOption, ...options];

  return (
    <Controller
      name={name}
      control={control}
      render={({ field: { onChange, value } }) => {
        return (
          <div className="flex flex-col">
            <label htmlFor={name} className="mb-0.5 text-sm">
              {label}
              {isRequired && (
                <span className="text-rose-600 dark:text-rose-400">
                  &nbsp;*
                </span>
              )}
            </label>
            <div className="relative">
              <select
                id={name}
                name={name}
                onChange={onChange}
                value={value}
                disabled={isDisabled}
                className={twMerge(
                  INPUT_FIELD_STYLES,
                  'appearance-none hover:cursor-pointer',
                  cn({
                    'border-rose-600 dark:border-rose-400': !!errorMessage
                  })
                )}
              >
                {optionsWithPlaceholder.map((option: Option) => (
                  <option
                    key={option.value}
                    value={option.value}
                    className="text-red-300"
                  >
                    {option.label}
                  </option>
                ))}
              </select>
              <ChevronUpDownIcon
                className="text-primary/80 pointer-events-none absolute top-1/2 right-2 -translate-y-1/2 dark:text-white/80"
                aria-hidden="true"
              />
            </div>
            {errorMessage && (
              <span className={INPUT_FIELD_ERROR}>{errorMessage}</span>
            )}
          </div>
        );
      }}
    />
  );
};
