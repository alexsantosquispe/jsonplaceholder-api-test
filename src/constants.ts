import { Option } from './services/api.types';

export const THEME_KEY = 'theme';

export const ALERT_TIMEOUT_DEFAULT = 3000;

export const ALERT_TYPES = {
  SUCCESS: 'success',
  ERROR: 'error',
  INFO: 'info'
};

export const DIGITS_LENGTH = 3;

export const ORDER_ASC = 'asc';

export const SORT_BY_NAME = 'name';

export const SORT_BY_TASK_TITLE = 'title';

export const INITIAL_PAGE_NUMBER = 1;

export const LAST_PAGE_NUMBER = 10;

export const PAGE_SIZES: Option[] = [
  { value: 10, label: '10' },
  { value: 20, label: '20' },
  { value: 30, label: '30' },
  { value: 50, label: '50' }
];

export const DEFAULT_PAGE_SIZE: Option = PAGE_SIZES[0];

export const COLUMNS_MAP = [
  { key: 'id', label: 'Code', isSortEnabled: true },
  { key: 'user', label: 'User', isSortEnabled: false },
  { key: 'title', label: 'Title', isSortEnabled: true },
  { key: 'completed', label: 'Status', isSortEnabled: true }
];
