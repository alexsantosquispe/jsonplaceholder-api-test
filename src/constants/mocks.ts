import { Option } from '../services/api.types';

export const OptionsMock: Option[] = [
  { value: '1', label: 'Option 1' },
  { value: '2', label: 'Option 2' },
  { value: '3', label: 'Option 3' },
  { value: '4', label: 'Option 4' },
  { value: '5', label: 'Option 5' },
  { value: '6', label: 'Option 6' }
];

export const LargeOptionsMock: Option[] = [
  { value: '1', label: 'Option 1' },
  { value: '2', label: 'This is option 2' },
  { value: '3', label: 'This is option 3' },
  { value: '4', label: 'This is option 4' },
  { value: '5', label: 'This is option 5' }
];
