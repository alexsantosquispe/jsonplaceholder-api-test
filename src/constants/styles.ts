export const INPUT_FIELD_STYLES =
  'w-full rounded-lg border border-gray-200 px-4 py-2 text-sm dark:border-white/30 focus:ring-primary/60 focus:border-primary/60 focus:ring-1 focus:outline-none dark:focus:border-white/65 dark:focus:ring-white/65';

export const INPUT_FIELD_ERROR = 'text-xs text-rose-600 dark:text-rose-400';

export const STRIPPED_BG =
  'md:bg-[repeating-linear-gradient(135deg,_var(--pattern-fg)_0,_var(--pattern-fg)_1px,_transparent_0,_transparent_50%)] md:bg-[size:10px_10px] md:bg-fixed md:dark:bg-[repeating-linear-gradient(135deg,_var(--pattern-fg-dark)_0,_var(--pattern-fg-dark)_1px,_transparent_0,_transparent_50%)]';
