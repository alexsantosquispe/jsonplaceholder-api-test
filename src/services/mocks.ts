export const UsersMock = [
  {
    value: '1',
    label: 'Leanne Graham'
  },
  {
    value: '2',
    label: 'Ervin Howell'
  },
  {
    value: '3',
    label: 'Clementine Bauch'
  },
  {
    value: '4',
    label: 'Patricia Lebsack'
  },
  {
    value: '5',
    label: 'Chelsey Dietrich'
  },
  {
    value: '6',
    label: 'Mrs. Dennis Schulist'
  },
  {
    value: '7',
    label: 'Kurtis Weissnat'
  },
  {
    value: '8',
    label: 'Nicholas Runolfsdottir V'
  },
  {
    value: '9',
    label: 'Glenna Reichert'
  },
  {
    value: '10',
    label: 'Clementina DuBuque'
  }
];
