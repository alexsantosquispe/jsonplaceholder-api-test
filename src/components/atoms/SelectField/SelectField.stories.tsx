import type { Meta, StoryObj } from '@storybook/react';
import { Control, FieldValues, useForm } from 'react-hook-form';

import { SelectField } from './SelectField';
import { OptionsMock } from '../../../constants/mocks';
import { Wrapper } from '../../../utils/stories-utils';

const meta = {
  title: 'Atoms/SelectField',
  component: SelectField,
  parameters: {
    layout: 'centered'
  }
} satisfies Meta<typeof SelectField>;

export default meta;

type Story = StoryObj<typeof SelectField>;

export const Default: Story = () => {
  const { control } = useForm();

  return (
    <div className="flex w-full gap-4">
      <Wrapper className="w-full">
        <SelectField
          label="Select field default"
          name="select-1"
          control={control as Control<FieldValues>}
          options={OptionsMock}
        />
      </Wrapper>
      <Wrapper className="w-full">
        <SelectField
          label="Select field required"
          name="select-2"
          control={control as Control<FieldValues>}
          options={OptionsMock}
          isRequired
        />
      </Wrapper>
      <Wrapper className="w-full">
        <SelectField
          label="Select field disabled"
          name="select-3"
          control={control as Control<FieldValues>}
          options={OptionsMock}
          isDisabled
        />
      </Wrapper>
      <Wrapper className="w-full">
        <SelectField
          label="Select with error"
          name="select-4"
          control={control as Control<FieldValues>}
          options={OptionsMock}
          errorMessage="This field has an error"
        />
      </Wrapper>
    </div>
  );
};

Default.args = {};
