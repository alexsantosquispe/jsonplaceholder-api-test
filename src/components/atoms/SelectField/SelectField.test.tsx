import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { act } from 'react';
import { Control, FieldValues, useForm } from 'react-hook-form';

import { SelectField, SelectFieldProps } from './SelectField';
import { OptionsMock } from '../../../constants/mocks';

const SelectFieldWrapper = (props: Omit<SelectFieldProps, 'control'>) => {
  const { control } = useForm();
  return <SelectField {...props} control={control as Control<FieldValues>} />;
};

describe('SelectField', () => {
  const errorMessage = 'Field with error';
  const props = {
    label: 'Select field',
    name: 'options',
    options: OptionsMock
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render the component correctly', () => {
    let component = render(<SelectFieldWrapper {...props} />);
    expect(component).toMatchSnapshot();

    component = render(<SelectFieldWrapper {...props} isRequired />);
    expect(component).toMatchSnapshot();
  });

  it('should show required indicator when isRequired is true', () => {
    render(<SelectFieldWrapper {...props} isRequired />);

    expect(screen.getByText('*')).toBeInTheDocument();
  });

  it('should display option and select one', async () => {
    render(<SelectFieldWrapper {...props} />);

    const selectField = screen.getByRole('combobox');

    expect(selectField).toBeInTheDocument();

    await act(async () => {
      await userEvent.selectOptions(selectField, '3');
    });

    const selectedOption = screen.getByRole('option', {
      name: 'Option 3'
    }) as HTMLOptionElement;

    expect(selectedOption.selected).toBe(true);

    expect((selectField as HTMLSelectElement).value).toBe('3');
  });

  it('should the select field be disabled', () => {
    render(<SelectFieldWrapper {...props} isDisabled />);

    const selectField = screen.getByRole('combobox');

    expect(selectField).toBeInTheDocument();

    expect(selectField).toBeDisabled();
  });

  it('should display the error message and error styles', () => {
    render(<SelectFieldWrapper {...props} errorMessage={errorMessage} />);

    const selectField = screen.getByRole('combobox');

    expect(selectField).toBeInTheDocument();

    expect(selectField).toHaveClass('border-rose-600');

    expect(screen.getByText(errorMessage)).toBeInTheDocument();
  });
});
