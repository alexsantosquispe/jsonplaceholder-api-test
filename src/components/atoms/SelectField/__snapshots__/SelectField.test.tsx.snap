// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectField should render the component correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex w-full flex-col"
      >
        <label
          class="mb-0.5 text-sm"
          for="options"
        >
          Select field
        </label>
        <div
          class="relative"
        >
          <select
            class="w-full rounded-lg border border-gray-200 px-4 py-2 text-sm dark:border-white/30 focus:ring-primary/60 focus:border-primary/60 focus:ring-1 focus:outline-none dark:focus:border-white/65 dark:focus:ring-white/65 appearance-none hover:cursor-pointer"
            id="options"
            name="options"
          >
            <option
              class="text-red-300"
              value=""
            >
              Select an option
            </option>
            <option
              class="text-red-300"
              value="1"
            >
              Option 1
            </option>
            <option
              class="text-red-300"
              value="2"
            >
              Option 2
            </option>
            <option
              class="text-red-300"
              value="3"
            >
              Option 3
            </option>
            <option
              class="text-red-300"
              value="4"
            >
              Option 4
            </option>
            <option
              class="text-red-300"
              value="5"
            >
              Option 5
            </option>
            <option
              class="text-red-300"
              value="6"
            >
              Option 6
            </option>
          </select>
          <svg
            class="size-4 text-primary/80 pointer-events-none absolute top-1/2 right-2 -translate-y-1/2 dark:text-white/80"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.25 15 12 18.75 15.75 15m-7.5-6L12 5.25 15.75 9"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex w-full flex-col"
    >
      <label
        class="mb-0.5 text-sm"
        for="options"
      >
        Select field
      </label>
      <div
        class="relative"
      >
        <select
          class="w-full rounded-lg border border-gray-200 px-4 py-2 text-sm dark:border-white/30 focus:ring-primary/60 focus:border-primary/60 focus:ring-1 focus:outline-none dark:focus:border-white/65 dark:focus:ring-white/65 appearance-none hover:cursor-pointer"
          id="options"
          name="options"
        >
          <option
            class="text-red-300"
            value=""
          >
            Select an option
          </option>
          <option
            class="text-red-300"
            value="1"
          >
            Option 1
          </option>
          <option
            class="text-red-300"
            value="2"
          >
            Option 2
          </option>
          <option
            class="text-red-300"
            value="3"
          >
            Option 3
          </option>
          <option
            class="text-red-300"
            value="4"
          >
            Option 4
          </option>
          <option
            class="text-red-300"
            value="5"
          >
            Option 5
          </option>
          <option
            class="text-red-300"
            value="6"
          >
            Option 6
          </option>
        </select>
        <svg
          class="size-4 text-primary/80 pointer-events-none absolute top-1/2 right-2 -translate-y-1/2 dark:text-white/80"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.25 15 12 18.75 15.75 15m-7.5-6L12 5.25 15.75 9"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`SelectField should render the component correctly 2`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex w-full flex-col"
      >
        <label
          class="mb-0.5 text-sm"
          for="options"
        >
          Select field
        </label>
        <div
          class="relative"
        >
          <select
            class="w-full rounded-lg border border-gray-200 px-4 py-2 text-sm dark:border-white/30 focus:ring-primary/60 focus:border-primary/60 focus:ring-1 focus:outline-none dark:focus:border-white/65 dark:focus:ring-white/65 appearance-none hover:cursor-pointer"
            id="options"
            name="options"
          >
            <option
              class="text-red-300"
              value=""
            >
              Select an option
            </option>
            <option
              class="text-red-300"
              value="1"
            >
              Option 1
            </option>
            <option
              class="text-red-300"
              value="2"
            >
              Option 2
            </option>
            <option
              class="text-red-300"
              value="3"
            >
              Option 3
            </option>
            <option
              class="text-red-300"
              value="4"
            >
              Option 4
            </option>
            <option
              class="text-red-300"
              value="5"
            >
              Option 5
            </option>
            <option
              class="text-red-300"
              value="6"
            >
              Option 6
            </option>
          </select>
          <svg
            class="size-4 text-primary/80 pointer-events-none absolute top-1/2 right-2 -translate-y-1/2 dark:text-white/80"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.25 15 12 18.75 15.75 15m-7.5-6L12 5.25 15.75 9"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
      </div>
    </div>
    <div>
      <div
        class="flex w-full flex-col"
      >
        <label
          class="mb-0.5 text-sm"
          for="options"
        >
          Select field
          <span
            class="text-rose-600 dark:text-rose-400"
          >
            Â *
          </span>
        </label>
        <div
          class="relative"
        >
          <select
            class="w-full rounded-lg border border-gray-200 px-4 py-2 text-sm dark:border-white/30 focus:ring-primary/60 focus:border-primary/60 focus:ring-1 focus:outline-none dark:focus:border-white/65 dark:focus:ring-white/65 appearance-none hover:cursor-pointer"
            id="options"
            name="options"
          >
            <option
              class="text-red-300"
              value=""
            >
              Select an option
            </option>
            <option
              class="text-red-300"
              value="1"
            >
              Option 1
            </option>
            <option
              class="text-red-300"
              value="2"
            >
              Option 2
            </option>
            <option
              class="text-red-300"
              value="3"
            >
              Option 3
            </option>
            <option
              class="text-red-300"
              value="4"
            >
              Option 4
            </option>
            <option
              class="text-red-300"
              value="5"
            >
              Option 5
            </option>
            <option
              class="text-red-300"
              value="6"
            >
              Option 6
            </option>
          </select>
          <svg
            class="size-4 text-primary/80 pointer-events-none absolute top-1/2 right-2 -translate-y-1/2 dark:text-white/80"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.25 15 12 18.75 15.75 15m-7.5-6L12 5.25 15.75 9"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex w-full flex-col"
    >
      <label
        class="mb-0.5 text-sm"
        for="options"
      >
        Select field
        <span
          class="text-rose-600 dark:text-rose-400"
        >
          Â *
        </span>
      </label>
      <div
        class="relative"
      >
        <select
          class="w-full rounded-lg border border-gray-200 px-4 py-2 text-sm dark:border-white/30 focus:ring-primary/60 focus:border-primary/60 focus:ring-1 focus:outline-none dark:focus:border-white/65 dark:focus:ring-white/65 appearance-none hover:cursor-pointer"
          id="options"
          name="options"
        >
          <option
            class="text-red-300"
            value=""
          >
            Select an option
          </option>
          <option
            class="text-red-300"
            value="1"
          >
            Option 1
          </option>
          <option
            class="text-red-300"
            value="2"
          >
            Option 2
          </option>
          <option
            class="text-red-300"
            value="3"
          >
            Option 3
          </option>
          <option
            class="text-red-300"
            value="4"
          >
            Option 4
          </option>
          <option
            class="text-red-300"
            value="5"
          >
            Option 5
          </option>
          <option
            class="text-red-300"
            value="6"
          >
            Option 6
          </option>
        </select>
        <svg
          class="size-4 text-primary/80 pointer-events-none absolute top-1/2 right-2 -translate-y-1/2 dark:text-white/80"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.25 15 12 18.75 15.75 15m-7.5-6L12 5.25 15.75 9"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
